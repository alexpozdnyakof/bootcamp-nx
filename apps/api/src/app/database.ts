import { Database } from 'sqlite3'

const dbFileName =
	process.env.NODE_ENV === 'test' ? `database-test` : `database`

const database = new Database(`${__dirname}/${dbFileName}.sqlite`, err => {
	if (err) throw new Error(err.message)
})

function migrate() {
	console.log({ __dirname })

	database.exec(
		'CREATE TABLE IF NOT EXISTS projects (id INTEGER PRIMARY KEY AUTOINCREMENT,title VARCHAR(200) NOT NULL,description TEXT NULL)'
	)

	if (process.env.NODE_ENV === 'test') {
		console.log('test')
		addMigrationData()
	}
}

function addMigrationData() {
	console.log({ __dirname })
	database.exec(
		`INSERT OR REPLACE INTO projects VALUES (1, 'ホームページのリニューアル', '新しいフレッシュホームページの制作過程はこちら'), (2, 'チェックアウトフォームのリファクタリング', '新しいチェックアウト フォームでユーザー エクスペリエンスを向上'), (3, '実験プロジェクト', null)`
	)
}

export { database, migrate, addMigrationData }

// 	'CREATE TABLE IF NOT EXISTS projects (id INTEGER PRIMARY KEY AUTOINCREMENT,title VARCHAR(200) NOT NULL,description TEXT NULL)'
// `INSERT OR REPLACE INTO projects VALUES (1, 'ホームページのリニューアル', '新しいフレッシュホームページの制作過程はこちら'), (2, 'チェックアウトフォームのリファクタリング', '新しいチェックアウト フォームでユーザー エクスペリエンスを向上'), (3, '実験プロジェクト', null)`
