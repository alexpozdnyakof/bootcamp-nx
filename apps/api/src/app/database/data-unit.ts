type UniqueId = number

type UpdatedCreatedTime = {
	created: number
	updated: number
}

interface EntityPropsMap {
	task: {
		text: string
		done: boolean
	}
	task_list: {
		title: string
	}
	project: {
		title: string
	}
}

type EntityType = 'task' | 'task_list' | 'project'

type DataUnit<T extends EntityType> = {
	id: UniqueId
	parentId: UniqueId
	type: T
	properties: EntityPropsMap[T] & UpdatedCreatedTime
	children: Array<UniqueId>
}

type EntityUnit = DataUnit<'task'> | DataUnit<'task_list'> | DataUnit<'project'>

type AbstractStore = Array<EntityUnit>

type DataUnitDTO = Pick<
	DataUnit<keyof EntityPropsMap>,
	'parentId' | 'children' | 'type'
> & {
	properties: Omit<
		DataUnit<keyof EntityPropsMap>['properties'],
		'created' | 'updated'
	>
}

type DataUnitUpdateValue = Partial<
	Pick<DataUnitDTO, 'properties' | 'children' | 'parentId'>
> &
	Pick<EntityUnit, 'id'>

export {
	AbstractStore,
	EntityType,
	UniqueId,
	EntityUnit,
	DataUnitDTO,
	DataUnit,
	DataUnitUpdateValue,
	UpdatedCreatedTime,
}
