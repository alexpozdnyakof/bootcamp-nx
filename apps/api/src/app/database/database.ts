import { Database } from 'sqlite3'

import * as fs from 'fs'
import * as path from 'path'

const dbFileName = process.env.NODE_ENV === 'test' ? `test` : `database`

const database = new Database(`${__dirname}/${dbFileName}.sqlite`, err => {
	if (err) throw new Error(err.message)
})
const sqlUrl =
	process.env.NODE_ENV === 'test'
		? path.join(__dirname, '../../sql')
		: path.join(__dirname, '/sql')

function migrate() {
	console.log({ __dirname })

	database.exec(fs.readFileSync(sqlUrl.concat('/projects.sql')).toString())

	if (process.env.NODE_ENV === 'test') {
		console.log('test')
		seeds()
	}
}

function seeds() {
	console.log({ __dirname })
	database.exec(
		fs.readFileSync(sqlUrl.concat('/projects-migration.sql')).toString()
	)
}

export { database, migrate }

// 'CREATE TABLE IF NOT EXISTS projects (id INTEGER PRIMARY KEY AUTOINCREMENT,title VARCHAR(200) NOT NULL,description TEXT NULL)'
// `INSERT OR REPLACE INTO projects VALUES
// (1, 'ホームページのリニューアル',
// '新しいフレッシュホームページの制作過程はこちら'),
// (2, 'チェックアウトフォームのリファクタリング', '新しいチェックアウト フォームでユーザー エクスペリエンスを向上'), (3, '実験プロジェクト', null)`
