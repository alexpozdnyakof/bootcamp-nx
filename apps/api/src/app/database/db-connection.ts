import knex from 'knex'
import knexConfig from '../../knexfile'
const dbConnection = () => {
	const envConfig = knexConfig[process.env.NODE_ENV ?? 'development']
	return knex(envConfig)
}

export default dbConnection()
