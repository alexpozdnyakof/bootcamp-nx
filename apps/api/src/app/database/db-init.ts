import { Database } from 'sqlite3'
import { readFileSync, readdirSync } from 'node:fs'
import { join } from 'path'

function readSqlDirectory(subdir: string): Array<string> {
	const path = join(
		__dirname,
		process.env.NODE_ENV === 'test' ? '../../sql' : '/sql',
		subdir
	)
	return readdirSync(path).map(fileName =>
		readFileSync(join(path, '/', fileName)).toString()
	)
}

function migrate(dbInstance: Database) {
	const migrations = readSqlDirectory('/migrations')
	migrations.map(sqlString => dbInstance.exec(sqlString))
}

function seed(dbInstance: Database) {
	const seeds = readSqlDirectory('/seeds')
	seeds.map(sqlString => dbInstance.exec(sqlString))
}

export function InitSqlite() {
	const isTest = process.env.NODE_ENV === 'test'
	const dbpath = isTest ? ':memory:' : join(__dirname, '/database.sqlite')

	const database = new Database(dbpath, err => {
		if (err) throw new Error(err.message)
	})

	migrate(database)
	if (process.env.NODE_ENV === 'test') {
		seed(database)
	}

	return database
}

export default InitSqlite()
