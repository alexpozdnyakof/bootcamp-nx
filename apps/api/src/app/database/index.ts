export { default as DataModel } from './data-model'
export * from './data-unit'
export { default as Database } from './db'
