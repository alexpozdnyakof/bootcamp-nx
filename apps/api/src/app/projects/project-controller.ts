import { ApiProject, ApiTask, ApiTaskList } from '@bootcamp-nx/api-interfaces'
import { Response, Router } from 'express'
import { ResponseWithData } from '../response-types'
import { CreateTask } from '../task/task'
import { CreateTaskList } from '../tasklist'
import { TypedResponse } from '../typed-response'
import { CreateApiProject } from './api-project'
import { ProjectValue } from './project'
import ProjectModel from './project-repo'

const ProjectRouter = Router()
const ProjectRouterPrefix = 'project'

ProjectRouter.get(
	'/',
	async (req, res: TypedResponse<ResponseWithData<Array<ApiProject>>>) => {
		try {
			const { id } = req.user
			const result = await ProjectModel.GetAll(id)
			res.status(200).send({ code: 200, data: result })
		} catch (error) {
			res.sendStatus(404)
		}
	}
)

ProjectRouter.get(
	'/:id',
	async (req, res: Response<ApiProject | { message: string }>) => {
		const id = Number(req.params.id)
		try {
			const projectRow = await ProjectModel.GetOne(id)
			if (projectRow === undefined) throw new Error('Not Found')

			const ApiProject = CreateApiProject(projectRow)

			res.status(200).send(ApiProject)
		} catch (error) {
			res.status(404).send({ message: error.message })
		}
	}
)

ProjectRouter.get(
	'/:id/sections',
	async (req, res: Response<ApiProject | { message: string }>) => {
		const id = Number(req.params.id)
		try {
			const projectRow = await ProjectModel.GetOne(id)
			if (projectRow === undefined) throw new Error('Not Found')

			const ApiProject = CreateApiProject(projectRow)

			res.status(200).send(ApiProject)
		} catch (error) {
			res.status(404).send({ message: error.message })
		}
	}
)

ProjectRouter.get(
	'/:id/tasklists',
	async (req, res: Response<Array<ApiTaskList> | { message: string }>) => {
		const id = Number(req.params.id)
		try {
			const tasklistsRow = await ProjectModel.GetRelatedTasklists(id)
			const apiTaskLists = tasklistsRow.map(row => CreateTaskList(row))
			res.status(200).send(apiTaskLists)
		} catch (error) {
			res.status(400).send({ message: error.message })
		}
	}
)
ProjectRouter.get(
	'/:id/tasks',
	async (
		req,
		res: Response<
			Array<ApiTask & { tasklist_id: number }> | { message: string }
		>
	) => {
		const id = Number(req.params.id)
		try {
			const tasksRow = await ProjectModel.GetRelatedTasks(id)

			const apiTasks = tasksRow.map(({ tasklist_id, ...row }) =>
				Object.assign(CreateTask(row), { tasklist_id })
			)
			res.status(200).send(apiTasks)
		} catch (error) {
			res.status(400).send({ message: error.message })
		}
	}
)

ProjectRouter.post('/', async (req, res) => {
	try {
		const dto = ProjectValue.check(req.body)
		const result = await ProjectModel.Add(dto)
		res.status(201).send(result)
	} catch (error) {
		res.status(400).send({ message: error.message })
	}
})

ProjectRouter.delete('/:id', async (req, res) => {
	const id = Number(req.params.id)
	try {
		await ProjectModel.Delete(id)
		res.status(200).send()
	} catch (error) {
		res.status(400).send({ message: error.message })
	}
})

ProjectRouter.put('/:id', async (req, res) => {
	const id = Number(req.params.id)
	try {
		const dto = ProjectValue.check(req.body)
		await ProjectModel.Update(id, dto)
		res.status(201).send()
	} catch (error) {
		res.status(400).send({ message: error.message })
	}
})

export { ProjectRouter as ProjectController, ProjectRouterPrefix }
