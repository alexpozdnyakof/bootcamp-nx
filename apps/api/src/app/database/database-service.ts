import database from './database'

type DbResult<T> = {
	status: 'success' | 'failed'
	result: T
}

export function DatabaseService() {
	return Object.freeze({
		async All<T>(
			query: string,
			params?: Array<string | number | boolean>
		): Promise<T> {
			return new Promise((resolve, reject) => {
				database.all(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
		async Get<T>(
			query: string,
			params?: Array<string | number | boolean>
		): Promise<T> {
			return new Promise((resolve, reject) => {
				database.get(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
		async Run<T>(
			query: string,
			params?: Array<string | number | boolean>
		): Promise<T> {
			return new Promise((resolve, reject) => {
				database.run(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
	})
}

export default DatabaseService()
