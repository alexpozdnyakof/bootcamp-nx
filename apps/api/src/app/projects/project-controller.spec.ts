import express from 'express'
import request from 'supertest'
import { json } from 'body-parser'
import { ProjectController } from './project-controller'
import { database } from '../database'

describe('ProjectController', () => {
	const App = express()
	App.use(json())
	App.use('/', ProjectController)

	beforeAll(async () => {
		await database.migrate.up({
			name: '20221129145930_create_project_table.js',
		})
		await database.migrate.up({
			name: '20221129145937_create_tasklist_table.js',
		})
		await database.migrate.up({
			name: '20221130102828_create_tasklist_to_project.js',
		})
		await database.migrate.up({
			name: '20221202114516_create_task_to_tasklist.js',
		})
		await database.migrate.up({
			name: '20221129145851_create_task_table.js',
		})
		await database.seed.run({ specific: '01-project.js' })
		await database.seed.run({ specific: '02-tasklist.js' })
		await database.seed.run({ specific: '03-task.js' })
	})

	xit('should return all projects', async () => {
		const response = await request(App).get('/1')
		expect(response.status).toBe(200)
		expect(response.body).toMatchSnapshot()
	})

	xit('should return all related tasklists', async () => {
		const response = await request(App).get('/1/tasklists')
		expect(response.status).toBe(200)
		expect(response.body).toMatchSnapshot()
	})

	xit('should return all related tasks', async () => {
		const response = await request(App).get('/1/tasks')
		expect(response.status).toBe(200)
		expect(response.body).toMatchSnapshot()
	})

	xit('should return project with id 1', async () => {
		const response = await request(App).get('/1')
		expect(response.status).toBe(200)
		expect(response.body).toMatchSnapshot()
	})

	xit('should return 404 when get non-existing project', async () => {
		const response = await request(App).get('/10')
		expect(response.status).toBe(404)
		expect(response.body).toEqual({ message: 'Not Found' })
	})

	xit('should create new one project', async () => {
		const response = await request(App)
			.post('/')
			.set('Accept', 'application/json')
			.send({
				title: '新しい計画',
				description: '簡単な説明',
			})

		expect(response.status).toBe(201)
	})

	xit('should not create invalid project', async () => {
		const response = await request(App)
			.post('/')
			.set('Accept', 'application/json')
			.send({
				message: '簡単な説明',
			})

		expect(response.status).toBe(400)
		expect(response.body).toMatchSnapshot()
	})

	xit('should delete exist project', async () => {
		const response = await request(App).delete('/3')
		expect(response.status).toBe(200)
	})

	xit('should update project', async () => {
		const response = await request(App).put('/1').send({
			title: '新しい計画',
			description: '簡単な説明',
		})
		expect(response.status).toBe(201)
	})

	xit('should return 404 when update non-existing project ', async () => {
		const response = await request(App).put('/10').send({
			title: '新しい計画',
			description: '簡単な説明',
		})

		expect(response.status).toBe(400)
		expect(response.body).toEqual({ message: 'Not Found' })
	})

	xit('should return 400 for project invalid update', async () => {
		const response = await request(App).put('/10').send({
			message: '簡単な説明',
		})

		expect(response.status).toBe(400)
		expect(response.body).toMatchSnapshot()
	})
})
