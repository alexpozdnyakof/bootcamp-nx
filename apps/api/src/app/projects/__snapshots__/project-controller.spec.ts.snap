// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectController should not create invalid project 1`] = `
Object {
  "message": "Validation failed:
{
  \\"title\\": \\"Expected string, but was missing\\",
  \\"description\\": \\"Expected string, but was missing\\"
}.
Object should match { title: string; description: string; }",
}
`;

exports[`ProjectController should return 400 for project invalid update 1`] = `
Object {
  "message": "Validation failed:
{
  \\"title\\": \\"Expected string, but was missing\\",
  \\"description\\": \\"Expected string, but was missing\\"
}.
Object should match { title: string; description: string; }",
}
`;

exports[`ProjectController should return all projects 1`] = `
Object {
  "created": "2022-11-29 15:31:37",
  "description": "新しいフレッシュホームページの制作過程はこちら",
  "id": 1,
  "owner_id": 1,
  "title": "ホームページのリニューアル",
  "updated": "2022-11-29 15:31:37",
}
`;

exports[`ProjectController should return all related tasklists 1`] = `
Array [
  Object {
    "created": "2022-11-29 15:31:37",
    "description": "プレリリースジョブのみ",
    "id": 1,
    "title": "すべてのフロントエンド タスク",
    "type": "task_list",
    "updated": "2022-11-29 15:31:37",
  },
  Object {
    "created": "2022-11-29 15:31:37",
    "description": "プレリリースジョブのみ",
    "id": 2,
    "title": "すべてのバックエンド タスク",
    "type": "task_list",
    "updated": "2022-11-29 15:31:37",
  },
]
`;

exports[`ProjectController should return all related tasks 1`] = `
Array [
  Object {
    "created": "2022-11-29 15:31:37",
    "done": false,
    "id": 1,
    "tasklist_id": 1,
    "title": "血液レポートのグラフが空白になっている",
    "type": "task",
    "updated": "2022-11-29 15:31:37",
  },
  Object {
    "created": "2022-11-29 15:31:37",
    "done": true,
    "id": 2,
    "tasklist_id": 1,
    "title": "無効にする|| ユーザーがアカウントを無効にできない",
    "type": "task",
    "updated": "2022-11-29 15:31:37",
  },
  Object {
    "created": "2022-11-29 15:31:37",
    "done": false,
    "id": 3,
    "tasklist_id": 2,
    "title": "プロフィール、プロフィールの編集、ポップアップ",
    "type": "task",
    "updated": "2022-11-29 15:31:37",
  },
]
`;

exports[`ProjectController should return project with id 1 1`] = `
Object {
  "created": "2022-11-29 15:31:37",
  "description": "新しいフレッシュホームページの制作過程はこちら",
  "id": 1,
  "owner_id": 1,
  "title": "ホームページのリニューアル",
  "type": "project",
  "updated": "2022-11-29 15:31:37",
}
`;
